envi/memory.py

```diff
@@ -525,31 +525,78 @@ def getMemoryMap(self, va):
     def getMemoryMaps(self):
         return [mmap for mva, mmaxva, mmap, mbytes in self._map_defs]
 
-    def readMemory(self, va, size):
+    def readMemory(self, va, size, origva=None):
         '''
         Read memory from maps stored in memory maps.
+
+        If the read crosses memory maps and fails on a later map, the exception
+        will show the details of the last map/failure, but should include the
+        original va (not the size).
         '''
         for mva, mmaxva, mmap, mbytes in self._map_defs:
             if mva <= va < mmaxva:
                 mva, msize, mperms, mfname = mmap
                 if not mperms & MM_READ:
-                    raise envi.SegmentationViolation(va)
+                    msg = "Bad Memory Read (no READ permission): %s: %s" % (hex(va), hex(size))
+                    if origva:
```

Comment:
```suggestion
                    if origva is not None:
```

because the VA 0x0 is valid but falsy, we should be explicit here about checking for `None`. same comment in subsequent changes.